Project4.md

```diff
@@ -212,62 +214,90 @@ using sufficient randomness, speed etc.
+      return render('login.html', ...)
+    # logged them in. set their login cookie and redirect to back to wherever they came from
+    authenticator = resp['resp']['authenticator']
+    response = HttpResponseRedirect(next)
+    response.set_cookie("auth", authenticator)
+    return response
+
+def create_listing(request):
+    auth = request.COOKIES.get('auth')
+    if not auth:
+      # handle user not logged in while trying to create a listing
+      return HttpResponseRedirect(reverse("login") + "?next=" + reverse("create_listing")
+    if request.method == 'GET':
+      return render("create_listing.html", ...)
+    f = create_listing_form(request.POST)
+    ...
+    resp = create_listing_exp_api(auth, ...)
+    if resp and not resp['ok']:
+        if resp['error'] == exp_srvc_errors.E_UNKNOWN_AUTH:
+            # exp service reports invalid authenticator -- treat like user not logged in
+            return HttpResponseRedirect(reverse("login") + "?next=" + reverse("create_listing")
+     ...
+     return render("create_listing_success.html", ...)
+```
+
 Note, the pattern of passing a URL argument of 'next' into the login
 page to specify where the user should be redirected upon succesfull login.
 The create_listing view can use this when it finds that the current user is not
 logged in. They can be redirected to the login page with next set to the url
 for the create_listing view. Then when they complete logging in, the login view
-will redirect them back to the create_listing.
+will redirect them back to the create_listing. The url will look something like this
+/account/login/?next=CreateListing
 
 Also notice the pattern of each view handling both rendering via GET and POST
 requests. The form for logging in or creating a listing will initially render
 via a GET and then be POST'ed upon submission. The POST handler will either error out
 and return the same rendered form (plus some helpful errors) or will succeed and return
 the user to wherever they're supposed to go next.
 
-As a further side note, you may wish to use Django's session middleware
-in your web front end for storing and retrieving authenticators
-instead of using cookies directly. It's fine if you want to do it this way.
+### Some Django best practices (Optional) ###
+
+#### Python decorator ####
+Suppose later your site has more services like create_list that requires the user 
+to be authenticated. One option is to rewrite the authenticating code for each 
+of the views. This violates the software engineering principle of DRY (Don't repeat yourself).
+Don't worry, python has the idea of nested functions, which in turn powers the idea of a decorator.
+You can think of decorator as an on-the-fly modification to a function. In this case you may consider
+creating a decorator to authenticate the user. Python provides the syntax sugar to attach a decorator 
+to a fuction as follows,
+
+```PYTHON
+@login_required
```

Comment:
Perhaps we should complete the example by showing a definition of login_required?